File: TypeScript/src/rock-paper-scissors.ts
class Player {
  constructor(private move: string) {
    this.move = move;
  }

  getMove() {
    return this.move;
  }
}

class RulesEngine {

  private rules = {
    scissors: "paper",
    paper: "rock",
    rock: "scissors",
  };

  getWinner(player1: Player, player2: Player): string {
    const player1Move = player1.getMove();
    const player2Move = player2.getMove();

    if (player1Move === player2Move) {
      return "Draw";
    }

    if (this.rules[player1Move] === player2Move) {
      return `Player 1 wins: ${player1Move} beats ${player2Move}`;
    }
    return `Player 2 wins: ${player2Move} beats ${player1Move}`;
  }
}

export class RockPaperScissors {
  private player1: Player;
  private player2: Player;

  constructor(private readonly input: string) {
    this.player1 = new Player(this.input.split(" ")[0]);
    this.player2 = new Player(this.input.split(" ")[1]);
  }

  play(): String {
    const engine = new RulesEngine();
    return engine.getWinner(this.player1, this.player2);
  }
}

// André Carvalho
// Pedro Alves
// André Coelho
// Brandão
// Anna
// José
// Solomiya

File: Kotlin/src/main/kotlin/RockPaperScissors.kt
class RockPaperScissors(private val input: String) {
    fun play(): String {
        val players = input.split(" ");
        val winner = getWinner(players);
        if (winner == -1) return "It's a draw!"
        return "Player ${winner+1} wins: ${players[winner]} beats ${players[((winner + 1) % 2)]}"
    }

    fun getWinner(players: List<String>): Int {
        if (players[0].equals(players[1], true)) return -1
        if (players[0].equals("rock") && players[1].equals("scissors")) return 0        
        if (players[0].equals("scissors") && players[1].equals("paper")) return 0        
        if (players[0].equals("paper") && players[1].equals("rock")) return 0
        return 1
    }
}